core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -144,288 +144,181 @@ public final BasicInvocationStage visitClearCommand(InvocationContext ctx, Clear
+         // Therefore, if we got here it means that we have lower topology than some other nodes and we can wait
+         // for it in StateTransferInterceptor and retry the read later.
+         // TODO: These situations won't happen as soon as we'll implement 4-phase topology change in ISPN-5021
+         throw new OutdatedTopologyException("Did not get any successful response, got " + responses);
       });
    }
 
-   protected Map<Object, InternalCacheEntry> retrieveFromRemoteSources(Set<?> requestedKeys, InvocationContext ctx, long flagsBitSet) throws Throwable {
-      GlobalTransaction gtx = ctx.isInTxScope() ? ((TxInvocationContext) ctx).getGlobalTransaction() : null;
-      CacheTopology cacheTopology = stateTransferManager.getCacheTopology();
-      ConsistentHash ch = cacheTopology.getReadConsistentHash();
+   protected final BasicInvocationStage handleNonTxWriteCommand(InvocationContext ctx, AbstractDataWriteCommand command)
+         throws Throwable {
+      Object key = command.getKey();
+      CacheEntry entry = ctx.lookupEntry(key);
 
-      Map<Address, List<Object>> ownerKeys = new HashMap<>();
-      for (Object key : requestedKeys) {
-         Address owner = ch.locatePrimaryOwner(key);
-         List<Object> requestedKeysFromNode = ownerKeys.get(owner);
-         if (requestedKeysFromNode == null) {
-            ownerKeys.put(owner, requestedKeysFromNode = new ArrayList<>());
+      if (isLocalModeForced(command)) {
+         if (entry == null) {
+            entryFactory.wrapExternalEntry(ctx, key, null, true);
          }
-         requestedKeysFromNode.add(key);
-      }
-
-      Map<Address, ReplicableCommand> commands = new HashMap<>();
-      for (Map.Entry<Address, List<Object>> entry : ownerKeys.entrySet()) {
-         List<Object> keys = entry.getValue();
-         ClusteredGetAllCommand remoteGetAll = cf.buildClusteredGetAllCommand(keys, flagsBitSet, gtx);
-         commands.put(entry.getKey(), remoteGetAll);
+         return invokeNext(ctx, command);
       }
 
-      RpcOptionsBuilder rpcOptionsBuilder = rpcManager.getRpcOptionsBuilder(
-            ResponseMode.SYNCHRONOUS_IGNORE_LEAVERS, DeliverOrder.NONE);
-      RpcOptions options = rpcOptionsBuilder.build();
-      // TODO Use multiple async invocations
-      Map<Address, Response> responses = rpcManager.invokeRemotely(commands, options);
-
-      Map<Object, InternalCacheEntry> entries = new HashMap<>();
-      for (Map.Entry<Address, Response> entry : responses.entrySet()) {
-         updateWithValues(((ClusteredGetAllCommand) commands.get(entry.getKey())).getKeys(),
-               entry.getValue(), entries);
+      if (entry == null) {
+         DistributionInfo info = new DistributionInfo(key, checkTopologyId(command).getWriteConsistentHash(), rpcManager.getAddress());
+         boolean load = shouldLoad(ctx, command, info);
+         if (info.isPrimary()) {
```

Comment:
It shouldn't, though I was about to handle any case correctly. I'll throw IllegalStateException instead.